File: Baekjoon/Class2/28.4949 균형잡힌 세상.java
import java.io.*;
import java.util.ArrayDeque;
import java.util.Deque;

class Main {
    private static boolean balancedWorld(String str) {
        Deque<Character> brackets = new ArrayDeque<Character>();
        for(char ch: str.toCharArray())
            if(ch == '(' || ch == '[') // 왼쪽 괄호 push
                brackets.push(ch);
            else if(ch == ')' || ch == ']') { // 오른쪽 괄호 pop
                if(brackets.isEmpty()) return false; // 스택 빔
                char top = brackets.pop();
                if(top != '(' && ch == ')') // 괄호 짝 안 맞음
                    return false;
                if(top != '[' && ch == ']')
                    return false;
            }
        return brackets.isEmpty(); // 남은 괄호 없으면 성공
    }

    public static void main(String[] args) throws IOException {
        String str;
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        while(!(str = br.readLine()).equals("."))
            if(balancedWorld(str)) sb.append("yes\n");
            else sb.append("no\n");
        System.out.print(sb.toString());
        br.close();
    }
}

File: Baekjoon/Class2/28.4949 균형잡힌 세상.cpp
#include <iostream>
#include <string>
#include <stack>
using namespace std;

bool balanced_world(const string& str) {
    stack<char> brackets;
    for (char ch : str)
        if (ch == '(' || ch == '[') // 왼쪽 괄호 push
            brackets.push(ch);
        else if (ch == ')' || ch == ']') { // 오른쪽 괄호 pop
            if (brackets.empty()) return false; // 스택 빔
            char top = brackets.top();
            if (top != '(' && ch == ')') // 괄호 짝 안 맞음
                return false;
            if (top != '[' && ch == ']')
                return false;
            brackets.pop();
        }
    return brackets.empty(); // 남은 괄호 없으면 성공
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    string str;
    while (getline(cin, str) && str != ".")
        if (balanced_world(str)) cout << "yes\n";
        else cout << "no\n";
    return 0;
}